signer/signer.go

```diff
@@ -101,6 +104,62 @@ type Signature interface {
 // SignedFile is an []bytes that contains file data
 type SignedFile []byte
 
+// GetKeysAndRand parses a configuration to retrieve the private and public key
+// of a signer, as well as a RNG and a marshalled public key. It knows to handle
+// HSMs as needed, and thus removes that complexity from individual signers.
+func (cfg *Configuration) GetKeysAndRand() (priv crypto.PrivateKey, pub crypto.PublicKey, rng io.Reader, publicKey string, err error) {
+	priv, err = cfg.GetPrivateKey()
+	if err != nil {
+		return
+	}
+
+	var publicKeyBytes []byte
+	switch priv.(type) {
+	case *rsa.PrivateKey:
+		pub = priv.(*rsa.PrivateKey).Public()
+		publicKeyBytes, err = x509.MarshalPKIXPublicKey(&priv.(*rsa.PrivateKey).PublicKey)
+		if err != nil {
+			err = errors.Wrap(err, "failed to asn1 marshal rsa public key")
+			return
+		}
+		rng = rand.Reader
+
+	case *ecdsa.PrivateKey:
+		pub = priv.(*ecdsa.PrivateKey).Public()
+		publicKeyBytes, err = x509.MarshalPKIXPublicKey(&priv.(*ecdsa.PrivateKey).PublicKey)
+		if err != nil {
+			err = errors.Wrap(err, "failed to asn1 marshal ecdsa public key")
+			return
+		}
+		rng = rand.Reader
+
+	case *crypto11.PKCS11PrivateKeyECDSA:
+		pub = priv.(*crypto11.PKCS11PrivateKeyECDSA).Public()
+		publicKeyBytes, err = x509.MarshalPKIXPublicKey(priv.(*crypto11.PKCS11PrivateKeyECDSA).PubKey.(*ecdsa.PublicKey))
+		if err != nil {
+			err = errors.Wrap(err, "failed to asn1 marshal crypto11 ecdsa public key")
+			return
+		}
+		rng = new(crypto11.PKCS11RandReader)
+
+	case *crypto11.PKCS11PrivateKeyRSA:
+		pub = priv.(*crypto11.PKCS11PrivateKeyRSA).Public()
+		publicKeyBytes, err = x509.MarshalPKIXPublicKey(priv.(*crypto11.PKCS11PrivateKeyRSA).PubKey.(*rsa.PublicKey))
+		if err != nil {
+			err = errors.Wrap(err, "failed to asn1 marshal crypto11 rsa public key")
+			return
+		}
+		rng = new(crypto11.PKCS11RandReader)
```

Comment:
I don't have a strong opinion. It might be ok to use the hsm's RNG for all signers is it's available, but like you said we might just want to move everything to the hsm anyway. Happy to go either way.

Reply:
:+1: I say we punt on it; there's no rush. We can land this, migrate more signer types to the HSM, and check back in a few months or lazily split the RNG code out if we have a signer that requires it.